states/games/snake.mod.py

```diff
@@ -11,9 +9,8 @@
 
 
 class Player:
-    def __init__(self, conn, addr, state):
```

Comment:
Ik zou het volgende willen voorstellen:

- deze Player class renamen naar BasePlayer, en deze als base class gebruiken voor alle game specifieke Player classes
- Elke state (game) heeft een property player_class met daarin de class die gebruikt moet worden voor die game, gebaseerd op de BasePlayer.
- BasePlayer heeft events (callsbacks) voor on_keypress, on_keydown/up, blabla etc.
- Als een player connect, wordt in de Socket class een nieuwe instance van de Player class voor die game gemaakt (kijkt dus naar player_class), bijgehouden met het 'sid'. Vervolgens wordt op de state (de game dus) player_joined(player) aangeroepen. De player parameter is dan een instance van de Player class specifiek voor die game (dus PongPlayer, PacmanPlayer, etc.)
- Als een event binnenkomt wordt in Socket de juiste player erbij gezocht dmv. de sid, en het juiste event aangeroepen. 

Hiermee hoef je in een game dus alleen een override te maken van BasePlayer en in player_joined iets met die player te doen. Je hoeft dan niet meer in elke game zelf de events te binden en al het gedoe met flags en de meerdere 'listeners' kan ook weg.

Reply:
Er zijn hierna nog steeds meerdere listeners nodig voor on_join, on_press, on_release en on_leave deze laatste 3 moeten alleen naar de player class verplaatst worden. 
Lijkt me verder wel een goed idee, alleen moet de Game ook mee gegeven kunnen worden aan de player, hoe wil je dit doen?

Reply:
waarom zijn er meerdere listeners nodig?

Over game die meegegeven moet: zie ook hieronder, de client moet kunnen aangeven welke game gestart moet worden. Daarmee heb je dan ook gelijk een instance van die game (en daarmee weer de juiste player_class)

Reply:
er zijn verschillende events, on_press, on_release, on_color. Deze events moeten nog steeds geregistreerd worden, alleen dan in de Player class

Reply:
als socket zorgt dat de juiste event uitkomt bij de juiste player instance, dan zie ik niet waarom die meerdere listeners nog nodig zijn. Socket heeft gewoon een lijst met player instances, geindexeerd op sid. Zodra er een event binnekomt van de client wordt daar players[sid].on_press() aangeroepen, that's it. De game krijgt dan op de juiste Player instance die event binnen.

Reply:
Het leave event moet blijven toch?

Reply:
Of mag dat ook in de player gebeuren

Reply:
ja, er komt dan een player_left(player) in BaseState, die wordt aangeroepen zodra een player disconnect/leaved. De player parameter is weer de instance van player, net als bij player_joined. We zouden die event ook op de player kunnen doen trouwens. Misschien wel beide, kan wel handig zijn.

Reply:
Laat maar, kan met overrides

Reply:
Probleem, sommige states willen custom arguments geven aan de Player class, zie states/games/pong.mod.py. Ook zijn er veel states die de Player classes predefined willen hebben

Reply:
Het is mss handiger om de player classes zo te houden als ze nu zijn, bij een on_join(player) event kan de Game state altijd zelf de player class aanmaken met de socket als argument

Reply:
We kunnen die games vast wel aanpassen. Extra arguments kun je met een optionele **kwargs doen.

Reply:
De player classes worden aangemaakt in de socket class, deze heeft de arguments niet

Reply:
```
        self.players.append(Player(start_pos=0, bounds=(0, h)))
        self.players.append(Player(start_pos=w - 1, bounds=(0, h)))
        self.players.append(Player(start_pos=0, bounds=(0, 32), orientation="horizontal"))
        self.players.append(Player(start_pos=0, bounds=(32, 64), orientation="horizontal"))
        self.players.append(Player(start_pos=0, bounds=(64, 96), orientation="horizontal"))
        self.players.append(Player(start_pos=h - 1, bounds=(0, 32), orientation="horizontal"))
        self.players.append(Player(start_pos=h - 1, bounds=(32, 64), orientation="horizontal"))
        self.players.append(Player(start_pos=h - 1, bounds=(64, 96), orientation="horizontal"))
```
Je kunt hier toch gewoon een dict maken in de state met player nummer -> parameters, en die gebruiken zodra de player joined, met een volgnummertje. Beide problemen opgelost.